### ISSUE TITLE
Epic: link to source code / GitHub in API docs

### ISSUE THREAD

**Issue by aeddins-ibm** [Qiskit/documentation#454]
### Feedback

Before the API docs migrated to ibm.com, IIRC the doc pages often included links to the relevant Github source code. [edit: apparently it may not have been github, but some other hosted version of the source code?] These are useful for verifying details of how the code behaves. It would be nice to have those back if it's not too technically difficult.

An example workflow: 
A user needs to know something about something in the API, say SparsePauliOp. They google "SparsePauliOp" and get the doc page (usually the top result), which might already have the info they need. If it doesn't, then the Github link on that page means with 1 click they get to the relevant source code, where they can find out what they need to know.

Now with the new docs, if the doc page does not have the desired info, then the user will probably do a new Google search "github qiskit", select the second result for qiskit/qiskit, then in that page use the Github search box to search for "SparsePauliOp", pick the most promising search-result snippet, etc. Of course it's doable but it adds time/clicks to something a user/developer might do many times, so direct links from the docs would be nice to have.

**Comment by javabster** [comment]
Yeah I agree I think a GitHub link would be preferred, I'm guessing our link checker would be able to catch if any of the  links get out of sync? I'm imagining a case where a branch name changes (e.g when it goes from main to stable to a historical version number) or something and then the url no longer works

**Comment by Eric-Arellano** [comment]
> I'm guessing our link checker would be able to catch if any of the links get out of sync?

Yes, although only as a nightly cron job because external link checking was really slow and sometimes flaky.

>  I'm imagining a case where a branch name changes (e.g when it goes from main to stable to a historical version number) or something and then the url no longer works

The link would always be a permalink to the exact commit used when the docs were built. That's important because code can change between versions.

**Comment by Eric-Arellano** [comment]
Thanks again for the upvotes everyone! 

FYI I split this into three tasks so that we can more easily track how things are going:

- https://github.com/Qiskit/documentation/issues/518
- https://github.com/Qiskit/documentation/issues/517
- https://github.com/Qiskit/documentation/issues/519

The designers are going to start soon on figuring out a design. Let us know if anyone wants to try out https://github.com/Qiskit/documentation/issues/517! Otherwise, one of us core maintainers will try to get to it near the end of December or earlier January.

**Comment by javabster** [comment]
question - is the plan for this source code link to just be available for the api-ref pages? If possible I think it could be a nice addition to the non-api-ref pages as well, what do you think?

**Comment by Eric-Arellano** [comment]
> question - is the plan for this source code link to just be available for the api-ref pages?

Yes, this is a link to the original source code for API files.

> If possible I think it could be a nice addition to the non-api-ref pages as well, what do you think?

I agree, but this should be tracked by a separate issue/epic. The visual design and the implementation look quite different. With API docs, the design is inline with the content, and we'll use Sphinx to determine what the exact URL will be.

**Comment by Eric-Arellano** [comment]
Great news! We got this working in https://github.com/Qiskit/documentation/pull/620 today and it's now live for historical API docs for [Runtime](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/0.14/qiskit_ibm_runtime.Estimator) 🎉

<img width="571" alt="Screenshot 2024-01-11 at 4 42 30 PM" src="https://github.com/Qiskit/documentation/assets/14852634/9abcc35a-b06a-44a8-a8fb-8e6668bdbf70">

Provider will be added in https://github.com/Qiskit/documentation/pull/624. The current version of Runtime will be added once https://github.com/Qiskit/qiskit-ibm-runtime/pull/1312 is merged. For Qiskit, we'll regenerate all the historical API versions and current docs once https://github.com/Qiskit/qiskit/pull/11548 is merged.

--

This initial implementation isn't perfect. https://github.com/Qiskit/documentation/issues/517 would make the links be more precise so it takes you to the exact lines, rather than the overall file; contributions appreciated on that! I think it should be doable for someone new to the project. 

https://github.com/Qiskit/documentation/issues/518 tracks making the visual design a little less awkward with how there isn't padding. But we didn't want to block this functionality on that once we realized yesterday this easier approach was possible.

--

Thank you everyone for the feedback on voting on this! It was really helpful for prioritization.

You may also want to subscribe to https://github.com/Qiskit/documentation/issues/479 for overall improvements to the API docs, one of the team's highest priorities.

**Comment by Eric-Arellano** [comment]
Update: Provider and all of Runtime are live as of yesterday 🎉

We've been working the past few days to get Qiskit source code links live. We've hit two road blocks but are making progress on fixing them:

1. For Qiskit 0.44 and earlier, the source code files refer to multiple GitHub repositories because Qiskit was a "metapackage" comprised of several smaller projects. This has required a few refactors like https://github.com/Qiskit/documentation/pull/644
2. For Qiskit 0.45, the source code links work. But a recent change to how we render math expressions is blocking us from deploying those docs :/

**Comment by Eric-Arellano** [comment]
This is now live for every version of the API docs 🚀 We used our link checker to validate that all of the URLs to GitHub work.

See the comment above https://github.com/Qiskit/documentation/issues/454#issuecomment-1888021357 for possible follow up improvements. 

Thank you everyone for voting on this issue!
